drivers/hmis/app/jobs/hmis/activity_log_processor_job.rb

```diff
@@ -0,0 +1,153 @@
+                  e_t[:data_source_id].eq(cas_t[:data_source_id]),
+                ].inject(&:and),
+              ),
+            ),
+          ).
+          pluck(cas_t[:id], e_t[:id]).to_h
+
+        enrollment_ids = assessment_ids.map { |assessment_id| enrollment_id_map[assessment_id&.to_i] }
+        RESOLVE_ENROLLMENT_IDS.call(enrollment_ids)
+      },
+      'Client' => ->(ids) {
+        ids.map { |id| { client_ids: [id.to_i] } }
+      },
+      'Enrollment' => RESOLVE_ENROLLMENT_IDS,
+      'EnrollmentSummary' => RESOLVE_ENROLLMENT_IDS,
+    }.freeze
+
+    def process_records(scope)
+      total_processed = 0
+      scope.find_in_batches do |batch|
+        # entity references in this batch
+        refs_by_type = batch.
+          flat_map { |record| record.resolved_fields.keys }. # ["graphql_name/id"]
+          map { |key| key.split('/', 2) }. # [[graphql_name, id]]
+          each_with_object({}) { |(key, value), obj| (obj[key] ||= []) << value } # {graphql_id => [id]}
+
+        # resolve entity references
+        # {"graphql_name/id" => {client_ids: [], enrollment_ids: []}}
+        resolved_refs = {}
+        RESOLVERS.each do |entity_type, resolver|
+          ids = refs_by_type[entity_type]&.uniq
+          next unless ids.present?
+
+          resolver.call(ids).each_with_index do |resolved, idx|
+            resolved_refs["#{entity_type}/#{ids[idx]}"] = resolved
+          end
+        end
+
+        # for each log record, build rows for insert from the resolved references
+        client_rows = []
+        enrollment_rows = []
+        batch.each do |log_record|
+          log_record.resolved_fields.keys.each do |key|
+            resolved = resolved_refs[key]
+            next unless resolved
+
+            resolved[:client_ids]&.compact&.each { |id| client_rows << [log_record.id, id] }
+            resolved[:enrollment_ids]&.compact&.each { |id| enrollment_rows << [log_record.id, id] }
+          end
+        end
+
+        # insert the rows and mark log_records as processed
+        scope.transaction do
+          # populate join tables
+          insert_rows(table_name: 'hmis_activity_logs_clients', references: 'client', rows: client_rows.uniq)
+          insert_rows(table_name: 'hmis_activity_logs_enrollments', references: 'enrollment', rows: enrollment_rows.uniq)
+          scope.where(id: batch.map(&:id)).update_all(processed_at: Time.current)
+        end
+        total_processed += batch.size
+      end
+      total_processed
+    end
+
+    def insert_rows(table_name:, references:, rows:)
+      return if rows.empty?
+
+      table = Arel::Table.new(table_name)
+      manager = Arel::InsertManager.new
+      manager.into(table)
+      manager.columns << table[:activity_log_id]
+      manager.columns << table[:"#{references}_id"]
+      manager.values = manager.create_values_list(rows)
+
+      connection.execute manager.to_sql
+    end
```

Comment:
This seems really complicated for a bulk insert, can we not use ActiveRecordImport? If this is significantly better performance, and we believe that's an expected issue, then fine, but I'd rather build up the array/hash of values and just call `klass.import!(hashes)`

Reply:
There's no active record class for these join tables; I assume ActiveRecordImport won't work for that. I guess I could add classes and a primary key to those tables but that feels like a lot just to support a different insert syntax. 

It seems like we use Arel pretty extensively in this app but perhaps there are parts of its API we want to avoid?

Reply:
Ah, I'd forgotten there were no classes to back the tables.  I guess if we don't have those or need them, this is fine.  We do make heavy use of Arel, but this is a new part of the API we haven't used, so it felt like something that we must be handling with ActiveRecord.  Overall our approach has been, use ActiveRecord until you can't, then use Arel until you can't or it becomes unwieldy, then use SQL.